src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilder.cs

```diff
@@ -100,38 +100,36 @@ public ValueTask Task
         public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
             where TAwaiter : ICriticalNotifyCompletion
             where TStateMachine : IAsyncStateMachine
-        {
-            _useBuilder = true;
-            _methodBuilder.AwaitUnsafeOnCompleted(ref awaiter, ref stateMachine);
-        }
+            => AsyncMethodBuilderCore.AwaitUnsafeOnCompleted(ref awaiter, ref stateMachine, ref m_task);
     }
 
     /// <summary>Represents a builder for asynchronous methods that returns a <see cref="ValueTask{TResult}"/>.</summary>
     /// <typeparam name="TResult">The type of the result.</typeparam>
     [StructLayout(LayoutKind.Auto)]
     public struct AsyncValueTaskMethodBuilder<TResult>
     {
-        /// <summary>The <see cref="AsyncTaskMethodBuilder{TResult}"/> to which most operations are delegated.</summary>
-        private AsyncTaskMethodBuilder<TResult> _methodBuilder; // mutable struct; do not make it readonly
+        private Task<TResult> m_task; // lazily-initialized: must not be readonly. Debugger depends on the exact name of this field.
         /// <summary>The result for this builder, if it's completed before any awaits occur.</summary>
         private TResult _result;
         /// <summary>true if <see cref="_result"/> contains the synchronous result for the async method; otherwise, false.</summary>
         private bool _haveResult;
-        /// <summary>true if the builder should be used for setting/getting the result; otherwise, false.</summary>
-        private bool _useBuilder;
```

Comment:
This isn't actually going to shrink the size of the builder, is it?  Assuming not, what are these changes helping?

Reply:
Depends on size of `TResult`

Reply:
What TResult could cause this to shrink as a result of removing this bool field?

Reply:
Actually can use a sentinel Task an drop `_haveResult `